File: leetcode/42-trapping-rain-water.py
# https://leetcode.com/problems/trapping-rain-water/
# Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water
# it can trap after raining.
# Constraints:
#   n == height.length
#   1 <= n <= 2 * 10^4
#   0 <= height[i] <= 10^5

from typing import List


class Solution:
    """
        Time Complexity: O(n)
        Auxiliary Space: O(1)
    """
    def trap(self, height: List[int]) -> int:
        out = 0

        max_left, max_right = height[0], height[-1]
        i, j = 1, len(height) - 2

        while i <= j:
            max_left = max(max_left, height[i])
            max_right = max(max_right, height[j])
            if max_left < max_right:
                out += max_left - height[i]
                i += 1
            else:
                out += max_right - height[j]
                j -= 1

        return out

File: leetcode/42-trapping-rain-water.go
// https://leetcode.com/problems/trapping-rain-water/
// Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water
// it can trap after raining.
// Constraints:
//    n == height.length
//    1 <= n <= 2 * 10^4
//    0 <= height[i] <= 10^5


func max(a int, b int) int {
    if a > b {
        return a
    }
    return b
}

//  Time Complexity: O(n)
//  Auxiliary Space: O(1)
func trap(height []int) int {
    out := 0
    i, j := 1, len(height) - 1
    max_i, max_j := height[0], height[j]

    for i <= j {
        max_i = max(max_i, height[i])
        max_j = max(max_j, height[j])

        if max_i < max_j {
            out += max_i - height[i]
            i += 1
        } else {
            out += max_j - height[j]
            j -= 1
        }
    }

    return out
}